src/jsp/jsp_cl.c

```diff
@@ -1614,3 +1630,154 @@ jsp_make_method_arglist (PARSER_CONTEXT * parser, PT_NODE * node_list)
+      data_type = data_type->data_type;
+    }
+
+  if (scale)
+    {
+      assert (data_type->type_enum == PT_TYPE_NUMERIC);
+      *scale = data_type->info.data_type.dec_precision;
+    }
+  else
+    {
+      assert (data_type->type_enum == PT_TYPE_CHAR || data_type->type_enum == PT_TYPE_VARCHAR);
+    }
+
+  *prec = data_type->info.data_type.precision;
+}
+
+static void
+jsp_stringify_data_type (char *buf, PT_TYPE_ENUM type_enum, PT_NODE * data_type)
+{
+  switch (type_enum)
+    {
+    case PT_TYPE_NONE:
+      strcpy (buf, "void");
+      return;
+    case PT_TYPE_RESULTSET:
+      strcpy (buf, "CURSOR");
+      return;
+    default:
+      ;				// do nothing
+    }
+
+  int db_type = pt_type_enum_to_db (type_enum);
+  strcpy (buf, db_get_name_of_db_type (db_type));
+
+  switch (db_type)
+    {
+    case DB_TYPE_NUMERIC:
+    case DB_TYPE_CHAR:
+    case DB_TYPE_VARCHAR:
+      {
+	char *prec_pos = buf + strlen (buf);
+	if (db_type == DB_TYPE_NUMERIC)
+	  {
+	    int prec, scale;
+	    jsp_get_prec_scale_from_data_type (&prec, &scale, data_type);
+	    sprintf (prec_pos, "(%d, %d)", prec, scale);
+	  }
+	else
+	  {
+	    // char or varchar
+	    int prec;
+	    jsp_get_prec_scale_from_data_type (&prec, NULL, data_type);
+	    sprintf (prec_pos, "(%d)", prec);
+	  }
+	break;
+      }
+    default:
+      ;				// do nothing
+    }
+}
+
+void
+jsp_parse_data_type_str (int *db_type, int *precision, int *scale, const char *data_type_str,
+			 const int data_type_str_len)
+{
+
+  // The way this function parses the data type string depends on
+  // the way the string was built in jsp_stringify_data_type()
+
+  bool has_precision, need_precision;
+  assert (db_type);
+
+  char buf[SP_DATA_TYPE_STR_LEN_MAX + 1];
+  memcpy (buf, data_type_str, data_type_str_len);
+  buf[data_type_str_len] = '\0';
+
+  char *lparen_pos = strchr (buf, '(');
+  if (lparen_pos)
+    {
+      has_precision = true;
+      *lparen_pos = '\0';
+    }
+  else
+    {
+      has_precision = false;
+    }
+
+  int ty;
+  if (strcmp (buf, "void") == 0)
+    {
+      ty = DB_TYPE_NULL;
+    }
+  else if (strcmp (buf, "CURSOR") == 0)
+    {
+      ty = DB_TYPE_RESULTSET;
+    }
+  else
+    {
+      ty = db_get_db_type_of_name (buf);	// it returns -1 when no match
+    }
+  assert (ty >= 0);
+  *db_type = ty;
+  need_precision = (ty == DB_TYPE_NUMERIC) || (ty == DB_TYPE_CHAR) || (ty == DB_TYPE_VARCHAR);
+
+  assert (has_precision == need_precision);
+
+  if (precision && scale)
+    {
+      if (need_precision)
+	{
+	  char *token = lparen_pos + 1, *delim;
+	  if (ty == DB_TYPE_NUMERIC)
+	    {
+	      delim = strchr (token, ',');
+	      *delim = '\0';
+	      *precision = atoi (token);
+
+	      token = delim + 2;	// 2: a space after the comma
+	      delim = strchr (token, ')');
+	      *delim = '\0';
+	      *scale = atoi (token);
+	    }
+	  else
+	    {
+	      // char or varchar
+	      delim = strchr (token, ')');
+	      *delim = '\0';
+	      *precision = atoi (token);
```

Comment:
```suggestion
	      *precision = atoi (token);
```
위와 같은 코드도 동일한 결과를 얻을 수 있습니다.

Reply:
그렇군요. 감사합니다.